modulemd/v2/modulemd-profile.c

```diff
@@ -0,0 +1,461 @@
+ * For more information on the license, see COPYING.
+ * For more information on free software, see <https://www.gnu.org/philosophy/free-sw.en.html>.
+ */
+
+#include <glib.h>
+#include <yaml.h>
+
+#include "modulemd-profile.h"
+#include "private/glib-extensions.h"
+#include "private/modulemd-profile-private.h"
+#include "private/modulemd-util.h"
+#include "private/modulemd-yaml.h"
+
+#define P_DEFAULT_STRING "__PROFILE_NAME_UNSET__"
+
+struct _ModulemdProfile
+{
+  GObject parent_instance;
+
+  gchar *name;
+  gchar *description;
+
+  GHashTable *rpms;
+};
+
+G_DEFINE_TYPE (ModulemdProfile, modulemd_profile, G_TYPE_OBJECT)
+
+enum
+{
+  PROP_0,
+
+  PROP_NAME,
+  PROP_DESCRIPTION,
+
+  N_PROPS
+};
+
+static GParamSpec *properties[N_PROPS];
+
+
+ModulemdProfile *
+modulemd_profile_new (const gchar *name)
+{
+  return g_object_new (MODULEMD_TYPE_PROFILE, "name", name, NULL);
+}
+
+
+ModulemdProfile *
+modulemd_profile_copy (ModulemdProfile *self)
+{
+  g_autoptr (ModulemdProfile) p = NULL;
+  g_return_val_if_fail (MODULEMD_IS_PROFILE (self), NULL);
+
+  p = modulemd_profile_new (modulemd_profile_get_name (self));
+
+  modulemd_profile_set_description (p,
+                                    modulemd_profile_get_description (self));
+
+  g_hash_table_unref (p->rpms);
+  p->rpms = g_hash_table_ref (self->rpms);
+
+  return g_steal_pointer (&p);
+}
+
+
+static void
+modulemd_profile_finalize (GObject *object)
+{
+  ModulemdProfile *self = (ModulemdProfile *)object;
+
+  g_clear_pointer (&self->name, g_free);
+  g_clear_pointer (&self->description, g_free);
+  g_clear_pointer (&self->rpms, g_hash_table_unref);
+
+  G_OBJECT_CLASS (modulemd_profile_parent_class)->finalize (object);
+}
+
+
+static void
+modulemd_profile_set_name (ModulemdProfile *self, const gchar *name)
+{
+  g_return_if_fail (MODULEMD_IS_PROFILE (self));
+  g_return_if_fail (name);
+  g_return_if_fail (g_strcmp0 (name, P_DEFAULT_STRING));
+
+  g_clear_pointer (&self->name, g_free);
+  self->name = g_strdup (name);
+
+  g_object_notify_by_pspec (G_OBJECT (self), properties[PROP_NAME]);
+}
+
+
+const gchar *
+modulemd_profile_get_name (ModulemdProfile *self)
+{
+  g_return_val_if_fail (MODULEMD_IS_PROFILE (self), NULL);
+
+  return self->name;
+}
+
+
+void
+modulemd_profile_set_description (ModulemdProfile *self,
+                                  const gchar *description)
+{
+  g_return_if_fail (MODULEMD_IS_PROFILE (self));
+
+  g_clear_pointer (&self->description, g_free);
+  self->description = g_strdup (description);
+
+  g_object_notify_by_pspec (G_OBJECT (self), properties[PROP_DESCRIPTION]);
+}
+
+
+const gchar *
+modulemd_profile_get_description (ModulemdProfile *self)
+{
+  g_return_val_if_fail (MODULEMD_IS_PROFILE (self), NULL);
+
+  return self->description;
+}
+
+
+void
+modulemd_profile_add_rpm (ModulemdProfile *self, const gchar *rpm)
+{
+  g_return_if_fail (MODULEMD_IS_PROFILE (self));
+  g_hash_table_insert (self->rpms, g_strdup (rpm), NULL);
```

Comment:
Please use [g_hash_table_add()](https://developer.gnome.org/glib/stable/glib-Hash-Tables.html#g-hash-table-add) instead. It's optimized for this precise usage.